presto-orc/src/main/java/com/facebook/presto/orc/DwrfDecryptorProvider.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc;
+
+import com.facebook.presto.orc.metadata.Encryption;
+import com.facebook.presto.orc.metadata.EncryptionGroup;
+import com.facebook.presto.orc.metadata.KeyProvider;
+import com.facebook.presto.orc.metadata.OrcType;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.slice.Slice;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+public class DwrfDecryptorProvider
+{
+    private Map<Integer, Integer> nodeToGroupMap;
+    private final Decryptor decryptor;
+
+    public DwrfDecryptorProvider(Encryption encryption, List<OrcType> types)
+    {
+        this.decryptor = getDecryptionLibrary(encryption.getKeyProvider());
+
+        List<EncryptionGroup> encryptionGroups = encryption.getEncryptionGroups();
+        ImmutableMap.Builder<Integer, Integer> nodeToGroupMapBuilder = ImmutableMap.builder();
+        for (int groupId = 0; groupId < encryptionGroups.size(); groupId++) {
+            // if group has key, use that.  otherwise use from stripe
+            for (Integer rootId : encryptionGroups.get(groupId).getNodes()) {
+                OrcType type = types.get(rootId);
+                for (int childId = rootId; childId < type.getFieldCount(); childId++) {
+                    nodeToGroupMapBuilder.put(childId, groupId);
+                }
+            }
+        }
+        this.nodeToGroupMap = nodeToGroupMapBuilder.build();
+    }
+
+    private Decryptor getDecryptionLibrary(KeyProvider keyProvider)
+    {
+        switch (keyProvider) {
+            case CRYPTO_SERVICE:
+                return getCryptoServiceDecryptor();
+            default:
+                return getUnknownDecryptor();
+        }
+    }
+
+    private Decryptor getCryptoServiceDecryptor()
+    {
+        // TODO: configure and pass through
+        return new DummyDecryptor();
+    }
+
+    private Decryptor getUnknownDecryptor()
+    {
+        // TODO: configure and pass through or throw an exception??
+        return new DummyDecryptor();
+    }
+
+    public Optional<DwrfDecryptor> createDecryptorIfEncrypted(int id, List<Slice> keyMetadatas)
```

Comment:
in theory, decrypter is stateless, and can be shared across all nodes in the same group